src/hmac.rs

```diff
@@ -0,0 +1,326 @@
+// Bitcoin Hashes Library
+// Written in 2018 by
+//   Andrew Poelstra <apoelstra@wpsoftware.net>
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+// This module is largely copied from the rust-crypto ripemd.rs file;
+// while rust-crypto is licensed under Apache, that file specifically
+// was written entirely by Andrew Poelstra, who is re-licensing its
+// contents here as CC0.
+
+//! # HMAC support
+
+use std::{fmt, io, ops};
+
+use {Error, Hash};
+
+/// A hash computed from a RFC 2104 HMAC. Parameterized by the underlying hash function.
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub struct Hmac<T: Hash>(T);
+
+/// Pair of underyling hash engines, used for the inner and outer hash of HMAC
+#[derive(Clone)]
+pub struct HmacEngine<T: Hash> {
+    iengine: T::Engine,
+    oengine: T::Engine,
+}
+
+impl<T: Hash> HmacEngine<T> {
+    /// Construct a new keyed HMAC with the given key
+    pub fn new(key: &[u8]) -> HmacEngine<T> {
+        debug_assert!(T::block_size() <= 128);
+
+        let mut ipad = [0x36u8; 128];
+        let mut opad = [0x5cu8; 128];
+        let mut ret = HmacEngine {
+            iengine: T::engine(),
+            oengine: T::engine(),
+        };
+
+        if key.len() > T::block_size() {
+            let hash = T::hash(key);
+            for (i, byte) in hash[..].iter().enumerate() {
```

Comment:
The for block could also be replaced by functional-style code, e.g.:
```rust
ipad.iter_mut().zip(&hash[..]).for_each(|(b_i, b_h)| *b_i ^= b_h);
opad.iter_mut().zip(&hash[..]).for_each(|(b_o, b_h)| *b_o ^= b_h);
```
I generally find it easier to reason about the correctness of such iterator chains than index access, but it's a rather minor nit. There is also a version that only needs one iteration instead of two, but it looks ugly (multiple zips).